core/src/main/java/io/atomix/core/collection/DistributedCollectionBuilder.java

```diff
@@ -30,4 +35,117 @@
+  }
+
+  /**
+   * Sets extra serializable types on the map.
+   *
+   * @param extraTypes the types to set
+   * @return the collection builder
+   */
+  @SuppressWarnings("unchecked")
+  public B withExtraTypes(Class<?>... extraTypes) {
+    config.setExtraTypes(Lists.newArrayList(extraTypes));
+    return (B) this;
+  }
+
+  /**
+   * Adds an extra serializable type to the map.
+   *
+   * @param extraType the type to add
+   * @return the collection builder
+   */
+  @SuppressWarnings("unchecked")
+  public B addExtraType(Class<?> extraType) {
+    config.addExtraType(extraType);
+    return (B) this;
+  }
+
+  /**
+   * Sets whether registration is required for serializable types.
+   *
+   * @return the collection configuration
+   */
+  @SuppressWarnings("unchecked")
+  public B withRegistrationRequired() {
+    return withRegistrationRequired(true);
+  }
+
+  /**
+   * Sets whether registration is required for serializable types.
+   *
+   * @param registrationRequired whether registration is required for serializable types
+   * @return the collection builder
+   */
+  @SuppressWarnings("unchecked")
+  public B withRegistrationRequired(boolean registrationRequired) {
+    config.setRegistrationRequired(registrationRequired);
+    return (B) this;
+  }
+
+  /**
+   * Sets whether compatible serialization is enabled.
+   *
+   * @return the collection builder
+   */
+  @SuppressWarnings("unchecked")
+  public B withCompatibleSerialization() {
+    return withCompatibleSerialization(true);
+  }
+
+  /**
+   * Sets whether compatible serialization is enabled.
+   *
+   * @param compatibleSerialization whether compatible serialization is enabled
+   * @return the collection builder
+   */
+  @SuppressWarnings("unchecked")
+  public B withCompatibleSerialization(boolean compatibleSerialization) {
+    config.setCompatibleSerialization(compatibleSerialization);
+    return (B) this;
+  }
+
+  /**
+   * Returns the protocol serializer.
+   *
+   * @return the protocol serializer
+   */
+  protected Serializer serializer() {
+    if (serializer == null) {
+      NamespaceConfig namespaceConfig = this.config.getNamespaceConfig();
+      if (namespaceConfig == null) {
+        namespaceConfig = new NamespaceConfig();
+      }
+
+      Namespace.Builder namespaceBuilder = Namespace.builder()
+          .register(Namespaces.BASIC)
+          .nextId(Namespaces.BEGIN_USER_CUSTOM_ID)
+          .register(new Namespace(namespaceConfig))
+          .nextId(Namespaces.BEGIN_USER_CUSTOM_ID + 100);
+
+      namespaceBuilder.setRegistrationRequired(config.isRegistrationRequired());
+      namespaceBuilder.setCompatible(config.isCompatibleSerialization());
+
+      if (config.getElementType() != null) {
+        namespaceBuilder.registerSubTypes(config.getElementType());
+      }
+      if (!config.getExtraTypes().isEmpty()) {
+        namespaceBuilder.register(config.getExtraTypes().toArray(new Class<?>[config.getExtraTypes().size()]));
```

Comment:
Better to use zero here instead of exact size.

Reply:
Using toArray(new T[0]) is usually the faster

See https://shipilev.net/blog/2016/arrays-wisdom-ancients/#_conclusion